### PR TITLE
feat(funcs): add semver functions

### PR DESCRIPTION
**Pull request by wuhuizuo**
- semver.Version: new a SemVer struct.
- semver.MatchConstraint: match in arg with semver constraint string.

refer to: https://github.com/Masterminds/sprig/blob/master/semver.go

Based on: #1828

### CODE REVIEW

**Comment by hairyhenderson** [comment]
Hi @wuhuizuo, thanks for submitting this, and sorry for the delay. Would it be possible for you to rebase and add some documentation for this?

**Comment by wuhuizuo** [comment]
> Hi @wuhuizuo, thanks for submitting this, and sorry for the delay. Would it be possible for you to rebase and add some documentation for this?

I update the documents, but I do not know how to update the markdown with make task:
```text
$ make docs/content/functions/semver.md
make: *** No rule to make target 'docs/content/functions/semver.md'.  Stop.
```

File: docs-src/content/functions/semver.yaml
Comment (hairyhenderson) [inline docs-src/content/functions/semver.yaml:1]: this file needs to be renamed `semver.yml`

File: docs-src/content/functions/semver.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+ns: semver
+preamble: |
+  These functions allow user you to parse a [semantic version](http://semver.org/) string or test it with constraint.
+  
+  It's implemented with https://github.com/Masterminds/semver.
```
Comment (hairyhenderson) [inline docs-src/content/functions/semver.yaml:5]:
```suggestion
  It's implemented with the https://github.com/Masterminds/semver library.
```

File: funcs/semver.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	ctx context.Context
+}
+
+// Semver -
+func (SemverFuncs) Semver(version string) (*semver.Version, error) {
+	return semver.NewVersion(version)
+}
+
+// MatchConstraint -
+func (SemverFuncs) MatchConstraint(constraint, in string) (bool, error) {
```
Comment (hairyhenderson) [inline funcs/semver.go:28]: Can we name this `CheckConstraint` instead so it's closer to the library's naming?

```suggestion
func (SemverFuncs) CheckConstraint(constraint, in string) (bool, error) {
```

File: funcs/semver_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			constraint: ">1.0.0",
+			in:         "va.b.c",
+			want:       false,
+			wantErr:    true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := SemverFuncs{
+				ctx: context.TODO(),
```
Comment (hairyhenderson) [inline funcs/semver_test.go:48]:
```suggestion
				ctx: context.Background(),
```

File: funcs/semver_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+			}
+			got, err := s.MatchConstraint(tt.constraint, tt.in)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("SemverFuncs.MatchConstraint() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("SemverFuncs.MatchConstraint() = %v, want %v", got, tt.want)
+			}
```
Comment (hairyhenderson) [inline funcs/semver_test.go:57]: I use `testify`'s `assert` package for testing in gomplate - please write these to match the same style (see other tests if you need a hint)

**Comment by hairyhenderson** [comment]
> I updated the documents, but I do not know how to update the markdown with make task:

Thanks @wuhuizuo - `make` is failing because you didn't correctly name the file - once you fix that it should run.

**Comment by wuhuizuo** [comment]
> > I updated the documents, but I do not know how to update the markdown with make task:
> 
> Thanks @wuhuizuo - `make` is failing because you didn't correctly name the file - once you fix that it should run.

Done, PTAL.

**Review by hairyhenderson — APPROVED** [review]
Thanks for your contribution @wuhuizuo! LGTM

**Pull request merged**